pkg/splunk/enterprise/afwscheduler.go

```diff
@@ -273,12 +280,177 @@ func getPhaseInfoByPhaseType(worker *PipelineWorker, phaseType enterpriseApi.App
+	updatePplnWorkerPhaseInfo(appDeployInfo, 0, enterpriseApi.AppPkgDownloadComplete)
+
+	scopedLog.Info("Finished downloading app")
+}
+
+// scheduleDownloads schedules the download workers to download app/s
+func (ppln *AppInstallPipeline) scheduleDownloads(pplnPhase *PipelinePhase, maxWorkers uint64, scheduleDownloadsWaiter *sync.WaitGroup) {
+	var activeWorkers uint64
+
+	scopedLog := log.WithName("scheduleDownloads")
+
+downloadWork:
+	for {
+		if activeWorkers <= maxWorkers {
+			select {
+			case _, channelOpen := <-ppln.sigTerm:
+				if !channelOpen {
+					scopedLog.Info("Received the termination request from the scheduler")
+					break downloadWork
+				}
+
+			case downloadWorker, ok := <-pplnPhase.msgChannel:
+				// if channel is closed, then just break from here as we have nothing to read
+				if !ok {
+					scopedLog.Info("msgChannel is closed by downloadPhaseManager, hence nothing to read.")
+					break downloadWork
+				}
+
+				// do not redownload the app if it is already downloaded
+				if isAppAlreadyDownloaded(downloadWorker) {
+					scopedLog.Info("app is already downloaded on operator pod, hence skipping it.", "appSrcName", downloadWorker.appSrcName, "appName", downloadWorker.appDeployInfo.AppName)
+					// update the state to be download complete
+					updatePplnWorkerPhaseInfo(downloadWorker.appDeployInfo, 0, enterpriseApi.AppPkgDownloadComplete)
+					continue
+				}
+
+				ppln.pplnMutex.Lock()
+				// do not proceed if we dont have enough disk space to download this app
+				if int64(ppln.availableDiskSpace-downloadWorker.appDeployInfo.Size) <= 0 {
+					// setting isActive to false here so that downloadPhaseManager can take care of it.
+					downloadWorker.isActive = false
+					ppln.pplnMutex.Unlock()
+					continue
+				}
+
+				// update the available disk space
+				ppln.availableDiskSpace = ppln.availableDiskSpace - downloadWorker.appDeployInfo.Size
+				ppln.pplnMutex.Unlock()
+
+				// increment the count in worker waitgroup
+				downloadWorker.waiter.Add(1)
+
+				// update the download state of app to be DownloadInProgress
+				updatePplnWorkerPhaseInfo(downloadWorker.appDeployInfo, downloadWorker.appDeployInfo.PhaseInfo.RetryCount, enterpriseApi.AppPkgDownloadInProgress)
+
+				appDeployInfo := downloadWorker.appDeployInfo
+
+				// create the sub-directories on the volume for downloading scoped apps
+				localPath, err := downloadWorker.createDownloadDirOnOperator()
+				if err != nil {
+					// increment the retry count and mark this app as download pending
+					updatePplnWorkerPhaseInfo(appDeployInfo, appDeployInfo.PhaseInfo.RetryCount+1, enterpriseApi.AppPkgDownloadPending)
+					continue
+				}
+
+				// get the S3ClientMgr instance
+				s3ClientMgr, _ := getS3ClientMgr(*downloadWorker.client, downloadWorker.cr, downloadWorker.afwConfig, downloadWorker.appSrcName)
+
+				// increment the number of active workers
+				activeWorkers++
```

Comment:
Do we need locks around `activeWorkers` here as well?

Reply:
Probably not here since `scheduleDownloads` is only a single thread

Reply:
I think its better if I use the counting semaphore method as is used by Subba in his PR, that way we will have a consistent code and we will not this field